### ISSUE TITLE
Question: Enviornment.ProcessorCount, Kubernetes/OpenShift and ThreadPool Min Threads

### ISSUE THREAD

**Issue by msallin** [dotnet/runtime#41586]
I read through a bunch of Issues and PR comments releated to this question but I'm unsure what the actual state is and whats planned, hence I'm creating this isssue.

I use .NET Core 3.1 (SDK version 3.1.401). The applications are deployed self contained as container to OpenShift.
By default the following CPU ressources are configured:
- Request: 100 milicores
- Limit: 500 milicores

This results in a Enviornment.ProcessorCount of 1 and hence in a ThreadPool min thread count of 1 as well.
This leads often to threadpool starvation sympthoms.

I know that the Enviornment.ProcessorCount can be adjusted by rising the limit but I'm unsure if this is the way to go with .NET Core 3.1 and also whats planned for .NET 5 (or later versions).

**Releated**
- https://github.com/dotnet/runtime/issues/11180
- https://github.com/dotnet/runtime/issues/13224
- https://github.com/dotnet/coreclr/pull/26153
- https://github.com/dotnet/coreclr/pull/23398
- https://github.com/dotnet/runtime/issues/622

**People active in these issues**
Maybe one of you can easily answer this question.
- @tmds 
- @stephentoub 
- @jkotas 
- @VSadov

**Comment by jkotas** [comment]
Could you please share more details about the thread starvation symptoms that you are seeing?

**Comment by msallin** [comment]
Sure. I might start with some details about the application.

The application consists of 5 Kafka consumers and several produces. A part of the consumed records get saved to a MSSQL database using EF Core. We are using [this](https://github.com/BEagle1984/silverback) library to communicate with Kafka . From external there are a bunch of incomming HTTP requests (not really worth to mention in terms of load. Checks for readiness, liveness and metrics and some requests to trigger reccuring actions).

There are three visible sympthoms:
- Kestrel HealthCheck warning about threadpool starvation
- Liveness & readiness checks sometimes fail because they timeout (5s)
- We use an EF Core interceptor to log potentially slow SQL commands, we get those warnings logged few times a day even if there was no SQL command slow (according to the MSSQL server), which is might due to the fact that a command has finished but couln't be immediatly picked up by a thread (async/await).

Our application code is straight forward and uses async await in a non-blocking manner. 
Currently we "solved" this by settings the minimum threads for the threadpool to 10.

**Comment by jkotas** [comment]
Were you able to identify the code that is blocking the threadpool thread when the heartbeat fails?

If this code is not under your control, it is fine to work around it be setting minimum thread pool threads. 

The threadpool does detect the starvation and it will inject additional thread pool threads to resolve it, but it takes some time for the threadpool to react and things may time out in the meantime as you have seen. We may be able to improve threadpool heuristics to react to this situation faster, but it is always going to be sub-optimal.

**Comment by jkotas** [comment]
cc @kouvel

**Comment by msallin** [comment]
> Were you able to identify the code that is blocking the thread pool thread when the heartbeat fails?

Not yet. I only see this occurring in production. I couldn't reproduce it locally nor on other environments. Unfortunately, we haven't yet the possibility to get dumps and/or traces from the container running in prod. We are working on it.

> The threadpool does detect the starvation and it will inject additional thread pool threads to resolve it, but it takes some time for the threadpool to react and things may time out in the meantime as you have seen. We may be able to improve threadpool heuristics to react to this situation faster, but it is always going to be sub-optimal.

Yep, this is true. The application never stays in this state and always recovers as soon as the thread pool created enough threads to handle the load.

**Comment by msallin** [comment]
I'm aware that the root of the problem most likely isn't (or at least not exclusive) the Enviornment.ProcessorCount of 1. However, my question remains because it affects also other libraries etc when the CPU count is 1.

Is the only and official way to let .NET know how much CPUs there really are to set the limit? Is there something planned in the future?

**Comment by jkotas** [comment]
The thread pool starvation problem is not limited to Environment.ProcessorCount of 1. It can happen even on e.g. 16 core machines and Environment.ProcessorCount returning 16 when there happens to be a sudden spike of 16+ tasks all blocking threadpool.

**Comment by GSPP** [comment]
MinThreads=1 seems extremely dangerous under most circumstances. For IO work, there need to be threads available essentially immediately or performance and safety are in danger.

I guess this would not show up when purely async IO is being used. Many real-world apps are not async-only (neither should they be since many real workloads do not derive any benefit from async IO). Accessing a (fast) database using sync IO is a totally fine choice in most cases.

.NET should fully support applications using synchronous IO. It is a bit against the trend these days but on a technical basis it often is the right choice. Sync IO should not be degraded to a niche scenario.

**Comment by tmds** [comment]
For many users .NET Core works fine with low CPU limits on Kubernetes.
We should try identify what is challenging about your workload.

It may be that some part of the application is still blocking ThreadPool Threads. This may be a 3rd party library you are using, or it could even be .NET Core itself.

Because your workload is bursty, the cpu allocation may not be enough to get all work done in a reasonable time.

If each request has a considerable cpu workload, it can also cause ThreadPool threads not becoming available in a reasonable time.

**Comment by msallin** [comment]
We had to focus on some other work. In the meantime, we switched to .NET 5.
Until now (beside too much noise in the logs) this behavior wasn't a problem. Now we are start working on applications where we can't tolerate this behavior anymore. Means we plan to invest time into troubleshooting, to find the root cause.

> It may be that some part of the application is still blocking ThreadPool Threads. This may be a 3rd party library you are using, or it could even be .NET Core itself.

We can't reproduce it outside of production i.e. Linux containers running on OpenShift. Any hints how to do this? 

> Because your workload is bursty, the cpu allocation may not be enough to get all work done in a reasonable time.
If each request has a considerable cpu workload, it can also cause ThreadPool threads not becoming available in a reasonable time.

That would mean that I should see the container hit the CPU limit or can't this be diagnosed like this? 
Anyway, we will try to just raise the cpu allocation and see if it changes something.

**Comment by davidfowl** [comment]
Are you tracking any metrics? If you're on k8s you might be able to deploy https://devblogs.microsoft.com/dotnet/introducing-dotnet-monitor/ and capture metrics for your application. There are a couple of thread pool counters that are interesting to collect. A future version of this tool will also allow configuring triggers to do things like collect dumps based on certain events or mertics 
cc @shirhatti

**Comment by msallin** [comment]
> Are you tracking any metrics? 

Yes, we tracking all metrics exposed as counters from System.Runtime and Microsoft.AspNetCore.Hosting. We use a sampling interval of two minutes, hold them in memory and let Prometheus scrape them all 2min as well. Works well and helps us to monitor the applications but for this case the sampling interval is too low.

>  If you're on k8s you might be able to deploy https://devblogs.microsoft.com/dotnet/introducing-dotnet-monitor/ and capture metrics for your application.

We did this already as an exercise to be prepared to get memory dumps, if a situation indicates this. I remember that tracing wasn't possible for some reason but if this is the best option, we will try to get it work.

> A future version of this tool will also allow configuring triggers to do things like collect dumps based on certain events or mertics

Thats gonna be extremely helpful because in our scenario we only see the behavior only a few times a day.

**Comment by msallin** [comment]
We discovered that our readiness health checks which are executed by K8s every 10s sometimes timeout (5s). We extended the timeout to 30s, to see if its a kind of deadlock. It isn't but the HealtCheck sometimes run for 12s. Behind the scenes the HealthCheck does a "SELECT 1". 

The Kestrel thread-pool starvation warning wasn't logged in this time and additionally, we increased the ThreadPoolMinThreads. The metrics show (at least for the 2min sample interval) that the threadpool never comes near the configured minimum of threads. We also increased the CPU limit, which has also no effect.

1. We profiled on the MSSQL Server to make sure, that this isn't the bottleneck (it comes at no surprise that the Profiler never shows a query which takes longer than 0ms).
2. To make sure that the network isn't the problem, we created a container with a sqlcmd which runs "SELECT 1" every 2s on the DB. We never saw taking a query longer than few ms.
3. We stripped down an application to the minimum which its necessary to run the EF health check (see code below).
Here we don't see that long times but still sometimes almost 1s. This application does nothing expect that the live and readiness check is polled by k8s all 10s (ThreadPoolMinThreads are set to 10).

For the last 7 days we saw ~17  queries which took >500ms.
The code of the application looks like this.

```
namespace HealthCheckTest
{
    public class Startup
    {
        public void ConfigureServices(IServiceCollection services)
        {
            const string connectionString = "Server = *, 1433; Database = *; User Id = *; Password = *;";

            services.AddDbContext<AppDbContext>(options =>
            {
                options.UseSqlServer(connectionString)
                    .LogTo(Console.WriteLine,
                        new[]
                        {
                            DbLoggerCategory.Database.Connection.Name,
                            DbLoggerCategory.Database.Command.Name
                        },
                        LogLevel.Debug,
                        DbContextLoggerOptions.SingleLine | DbContextLoggerOptions.LocalTime);
            });

            services
                .AddHealthChecks()
                .AddDbContextCheck<AppDbContext>(tags: new[] {"ready"});

        }

        public void Configure(IApplicationBuilder app)
        {
            app.UseRouting();
            app.UseEndpoints(endpoints =>
            {
                endpoints.MapHealthChecks("/health/ready", new HealthCheckOptions
                {
                    Predicate = check => check.Tags.Contains("ready")
                });
                endpoints.MapHealthChecks("/health/live", new HealthCheckOptions
                {
                    Predicate = _ => false
                });
            });
        }
    }

    public class AppDbContext : DbContext
    {
        public AppDbContext(DbContextOptions<AppDbContext> options)
            : base(options)
        {
        }
    }
}
```

**Comment by msallin** [comment]
Looks like the CPU time is being throttled by K8s, even we don't use the max according to the limit.

**Comment by msallin** [comment]
For the records (and anybody which maybe reads this issue): The CPU was heavily throttled but this doesn't seem to be the problem.

Now we have base images with batteries included (i.e., they contain the diagnostics tools https://github.com/dotnet/diagnostics/issues/1984).
That means we are able to trace in production. 

It will get very interesting. Because what I currently understand about our problem doesn't make sense at all.

- Those ~6s of time until a result of a awaited task gets processed. However, the thread pool has never reached its configured min threads and its also not growing, according to the metrics we collect all 2min. e.g. in one application we see a a max of 14 threads in the pool where the configured min threads would be 20.
- There is no correlation between higher load & when this happens.
- It has no influence on other activities of the application. The logs show that all other things going on continue as usual.
- These situations don't happen often (there are several million operations where we would detect this per day but it happens only ~100 times in this period)
- Have have no single Kestrel thread pool starvation warning in the logs

**Comment by davidfowl** [comment]
Do you have the dotnet counter information when this happened?

**Comment by msallin** [comment]
Yes and no. We are sampling the counters all 2min and scrape them with Prometheus. Thus, the resolution is low.
However, as we have now dotnet counters inside the container, I can sample the interesting counters every second.

Here is what we see. This is a HTTP call from K8s to the health check endpoint.
The warning is logged from a custom EF interceptor.

2021-02-11 11:07:21.736+00:00 level=INFO eventId=2 eventName=RequestFinished sourceContext="Microsoft.AspNetCore.Hosting.Diagnostics" message="Request finished HTTP/1.1 GET http://*:5000/health/ready - - - 200 - application/json;+charset=utf-8 3207.5906ms" elapsedMilliseconds="3207.5906" statusCode="200" contentType="application/json; charset=utf-8" contentLength="" protocol="HTTP/1.1" method="GET" scheme="http" host="10.127.67.3:5000" pathBase="" path="/health/ready" queryString="" spanId="a5748b6270afd443" traceId="bea0af4555a16244ab99d983f8901d0e" parentId="0000000000000000" connectionId="0HM6EEBSSMENE" requestId="0HM6EEBSSMENE:00000002" requestPath="/health/ready"

2021-02-11 11:07:21.735+00:00 level=WARN eventId=1000 eventName=PotentiallySlowDbCommand sourceContext="Post.Common.EntityFrameworkCore.Diagnostics.LogSlowDbCommandInterceptor" message="The execution of a DB-Command took longer than expected. Timeout: 15000ms, Time taken: **3207ms**, Method: ExecuteNonQuery, **Command: SELECT 1**" timeoutMilliseconds="15000" elapsedMilliseconds="3207" dbCommandMethod="ExecuteNonQuery" spanId="a5748b6270afd443" traceId="bea0af4555a16244ab99d983f8901d0e" parentId="0000000000000000" connectionId="0HM6EEBSSMENE" requestId="0HM6EEBSSMENE:00000002" requestPath="/health/ready" healthCheckName="AppDbContext"

2021-02-11 11:07:18.528+00:00 level=INFO eventId=1 eventName=RequestStarting sourceContext="Microsoft.AspNetCore.Hosting.Diagnostics" message="Request starting HTTP/1.1 GET http://*:5000/health/ready - -" protocol="HTTP/1.1" method="GET" contentType="" contentLength="" scheme="http" host="10.127.67.3:5000" pathBase="" path="/health/ready" queryString="" spanId="a5748b6270afd443" traceId="bea0af4555a16244ab99d983f8901d0e" parentId="0000000000000000" connectionId="0HM6EEBSSMENE" requestId="0HM6EEBSSMENE:00000002" requestPath="/health/ready"

The timestamps from the log lines are UTC. It is +1 for local time.

**Comment by msallin** [comment]
@davidfowl we observed the counters for a whole day (1s sampling interval). The thread pool threads never reached the maximum configured min threads. Hence, I conclude that its not a thread pool issue.

We also found some other interesting things like that it seems that the time spend is not when we receive the response but before even sending the request/query/command.  However, we are still unsure if its a infrastructure problem or something else. Hence, I created the following issue: https://github.com/dotnet/efcore/issues/24169

Because of our problem I tried to understand the role of the IO threads. But so far I wasn't able to understand how a starvation of those would manifest. Would it lead to the same symptoms?
